### PR TITLE
771 worst possible result

### PR DESCRIPTION
**Pull request by franchuterivera**
Created a function that based on the metric determines the worst cost for crash

Smac uses a MAX int defined as a custom constant, I am just picking a value given by NumPy which is already loaded

 I updated a test to make sure metrics whose optimum is zero are taken into account for coverage purposes.

### CODE REVIEW

File: autosklearn/evaluation/__init__.py [hunk 1]
```diff
@@ -35,13 +33,28 @@ def fit_predict_try_except_decorator(ta, queue, **kwargs):
...
                    'final_queue_element': True})
 
 
+def get_cost_of_crash(metric):
+
+    # For metrics like accuracy that are bounded to [0,1]
+    # assume 1 as the worst cost as autosklearn works with err
+    worst_possible_result = 1.0
+
+    if metric is not None:
```
Comment (mfeurer) [inline autosklearn/evaluation/__init__.py:49]: Hey, how can the metric be `None` here?

File: autosklearn/evaluation/__init__.py [hunk 2]
```diff
@@ -35,13 +33,28 @@ def fit_predict_try_except_decorator(ta, queue, **kwargs):
...
+        # Else, we use the make_scorer info to define the
+        # worst possible result
+        if metric._sign < 0:
+            worst_possible_result = np.iinfo(np.uint32).max
```
Comment (mfeurer) [inline autosklearn/evaluation/__init__.py:53]: As there every metric has the attribute `optimum`, I'm actually wondering whether we should add a new attribute `worst_possible_value` to each metric so that it's clearly defined (and only in one place, namely the metrics module) how they work. What do you think?

**Comment by franchuterivera** [comment]
- Make sure that a metric that inherits from scorer is passed (just a paranoia and in-code test)
- Now the metric could be none, when doing unit testing. I updated the automl class to handle a metric provided by the user. In the normal workflow, the estimators make sure that if the metric is not, accuracy for instance is used.
- added a comment to clarify  _loss functionality.
- The expected workflow:

1. A metric has optimum and worst results.
2. As autosklearn minimizes, a function get_worst_posible result translates the metric._worst_possible_result in a way the a minimization engine would understand that this is a worst possible result. Example: 
For accuracy, optimum=1 and worst=0. The worst possible error is optimum-worst=1-0=1. For
For logloss= optimum=0 and worst=MAXTINT= worst possible error is MAXINT

File: autosklearn/automl.py [hunk 3]
```diff
@@ -89,6 +89,7 @@ def __init__(self,
                  get_smac_object_callback=None,
                  smac_scenario_args=None,
                  logging_config=None,
+                 metric=None,
```
Comment (mfeurer) [inline autosklearn/automl.py:92]: I don't quite see why you pass the metric in here and do not use it later on.

File: test/test_metric/test_metrics.py [hunk 4]
```diff
@@ -116,7 +118,7 @@ def test_proba_scorer_binary(self):
         y_pred = [[1.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.0, 1.0]]
 
         scorer = autosklearn.metrics._ProbaScorer(
-            'accuracy', sklearn.metrics.log_loss, 0, 1, {})
+            'accuracy', sklearn.metrics.log_loss, 0, MAXINT, 1, {})
```
Comment (mfeurer) [inline test/test_metric/test_metrics.py:121]: Actually, as you're touching these lines, could you please change the names of the scorers?

File: autosklearn/automl.py (hunk 3 above)
Comment (franchuterivera) [reply autosklearn/automl.py:92]: To have it documented here:
We are adding the metric in the class instantiation because the automl object gets created directly during unit testing so that the dummy classifier can be tested.

Not needed during the normal auto-sklearn flow. Long term we would cleanup the fit/related methods because currently both the fit and instantiation takes the metric as argument.

**Comment by mfeurer** [comment]
Closes #771
